GA.java

```diff
@@ -1,26 +1,91 @@
-
+import java.util.Random;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
 public class GA extends Learning {
   Strategy[] population;
+	private static int NUM_POPULATION = 16;
+	private static int NUM_NEW_POPULATION = 42;
+
+
 	/**
 	 * Constructor
 	 */
 	public GA() {
-		
+		population = new Strategy[NUM_POPULATION];
+		if (!load()) {
+			Random rnd = new Random();
+			for (int i = 0; i < NUM_POPULATION; i++) {
+				double[] weights = new double[21];	
+				for (int j = 0; j < 21; j++) {
+					weights[j] = rnd.nextInt();
+				}
+				Strategy s = new Strategy(weights);
+				s.normalize();
+				population[i] = s;
+			}
+		}
+	}
+
+	class ScoredStrategy implements Comparable<ScoredStrategy> {
+		private Strategy s;
+		private double score;
+
+		public ScoredStrategy(Strategy s, double score) {
+			this.s = s;
+			this.score = score;
+		}
+
+		public int compareTo(ScoredStrategy s) {
+			if (s.score < this.score) {
+				return -1;
+			} else if (s.score == this.score) {
+				return 0;
+			} else {
+				return 1;
+			}
+		}
 	}
-	
 	/**
-	 * 
-	 * @param seed
+	 * Run some iterations of the genetic algorithm to refine the population.
+	 * @param iterations The number of iterations.
 	 */
-	public void run(int seed) {
-		
+	public void run(int iterations) {
+		Random rnd = new Random();
+    for (int i = 0; i < iterations; i++) {
+			List<Strategy> newPopulation = new ArrayList<Strategy>();
+			for (int j = 0; j < NUM_NEW_POPULATION/2; j++) {
+				int idx1 = rnd.nextInt(NUM_POPULATION);
+				int idx2 = rnd.nextInt(NUM_POPULATION);
+				Strategy[] newStrats = population[idx1].crossover(population[idx2]);
+				for (int k = 0; k < 2; k++) {
+					newStrats[k].mutate();
+					newStrats[k].normalize();
+					newPopulation.add(newStrats[k]);
+					newPopulation.add(newStrats[k]);
```

Comment:
Two strategies can be retrieved from one crossover (first part of A + second part of B, first part of B + second part of A)